### ISSUE TITLE
Add scrollbar to code editor

### ISSUE THREAD

**Issue by tkozybski** [staxrip/staxrip#567]
If you have large amount of (long) TextBoxes in Avisynth, the height of the Code Editor popup may exceed the monitor height, making it impossible to confirm changes. I have many various filters added with my notes about them so it takes a bunch of space.

Could scrollbar be added to Code Editor?

**Comment by Dendraspis** [comment]
Do you already use the hotfix EXE file from your last issue?

**Comment by tkozybski** [comment]
Yes

**Comment by Dendraspis** [comment]
Okay, to like last time simply replacing it (without any further attentions):

[StaxRip v2.1.8.4D_CEscroll.zip](https://github.com/staxrip/staxrip/files/6095136/StaxRip.v2.1.8.4D_CEscroll.zip)

**Comment by Dendraspis** [comment]
> the Save button is visible, but now the top is above the screen.

What do you mean?  🤔  Maybe it's a good idea to provide a screenshot - and opening a new issue, if there is an issue?!  🤔

**Comment by tkozybski** [comment]
It added scrollbars to individual boxes, but that still limits you before it gets too large. I thought about adding one scrollbar on the right for the whole thing instead of individual boxes. Also last box is now missing and is a big empty space? I'll provide a screenshot. On left is the old (yesterday's) version, right is the new one. That's how it looks on my 1440p monitor. I deleted the text because it doesn't matter.

[![scrollbar-issue.png](https://i.postimg.cc/Z5Jythbz/scrollbar-issue.png)](https://postimg.cc/bsFvkWSm)

**Comment by 44vince44** [comment]
@Dendraspis I tested it, I found all good, here are my comments:

- Scroll activates when many lines are typed in the filter 🆗 
- If many filter slots used and many lines inside each filter slot(VERY UNLIKELY), there could be a problem, so the solution is that the user uses "JOIN FILTERS" 🆗 very acceptable solution
- "the Save button is visible, but now the top is above the screen" I understand it, but could not reproduce it: the Code Editor window's top y coordinates should be the same as Staxrip Main Window, or a few pixels below, and NOT above, in order to avoid that (but as I said, could not reproduce it)

**Comment by Dendraspis** [comment]
@44vince44 
Do you think the vertical scrollbar is necessary or important?  🤔 

@tkozybski 
Because each section is separated, there can't be just one scrollbar for all.

**Comment by 44vince44** [comment]
> Do you think the vertical scrollbar is necessary or important? 🤔

Well yes, necessary because we have to ensure that there is a safety in case of "extreme cases". So now, it works like this:
- scroll appears when extreme number of lines are typed
- if extreme number of kines are typed AND many filter slots are in use, use FILTER JOIN.

I find it good as you just did it now.

**Comment by Dendraspis** [comment]
Okay, here is the last state:
- Vertical scrollbars appear when the content is longer than the height of the textbox
- *Joining filters* is refurbished - don't forget to reset the *Edit Menu*. 
- *Source* and *Rotation* filters are not merge anymore

[StaxRip v2.1.8.4D_CEscroll2.zip](https://github.com/staxrip/staxrip/files/6095251/StaxRip.v2.1.8.4D_CEscroll2.zip)

**Comment by 44vince44** [comment]
Amazing new joining options !!!

> Source and Rotation filters are not merge anymore

Agreed !
Well done and Thanks !!! 👍🏼

**Comment by tkozybski** [comment]
Again, in these new versions last filter is missing. They are not joined and it takes more space than in 2.1.8.0 (because of those border lines?).

[![scrollbar-issue-2.png](https://i.postimg.cc/Zn8KVjMx/scrollbar-issue-2.png)](https://postimg.cc/XBvbvK3r)

**Comment by 44vince44** [comment]
did you reset the menu of the code editor ?

**Comment by tkozybski** [comment]
> 
> 
> did you reset the menu of the code editor ?

I did reset the edit menu in Tools->Edit Menu.

When I deleted one of the filters the last one appeared, implying that there is some kind of limit now.

**Comment by 44vince44** [comment]
What you reset is not the Code Editor's menu.

Reset the Code Editor's menu like this:

open Code Editor
Right click in any filter slot
Select Edit Menu
Tools > Reset Everything

**Comment by tkozybski** [comment]
Didn't change a thing.

**Comment by 44vince44** [comment]
I can't reproduce what you say....

**Comment by tkozybski** [comment]
> 
> 
> I can't reproduce what you say....

Try adding several (I've got 5) filters with 10+ lines in each of them.

**Comment by 44vince44** [comment]
I know what's happening, you are using a previously saved project or template.
This new feature will work correctly on new project and new template.

**Comment by Dendraspis** [comment]
> imho: resize and crop must not merge either because they are special and linked to the main UI. they must be treated as Source and Rotation.

You're absolutely right!  😁 
[StaxRip v2.1.8.4D_CEscroll3.zip](https://github.com/staxrip/staxrip/files/6095340/StaxRip.v2.1.8.4D_CEscroll3.zip)

@tkozybski 
Space is limited, so in order to prevent issues with *too many filters*, you have to merge them to decrese the number of showing filters.

**Comment by 44vince44** [comment]
I found something, will try to reproduce it, wait...

**Comment by 44vince44** [comment]
Yes there is a problem, filters did not get deleted after merge, then filters got deleted when joined 🤣 
here is how to reproduce:
1) delete your settings, so you open defaults (keep it as is, AVS)
2) you have 
- Source
- Crop
- Field
- Noise
- Resize
Open code Editor

3) in Noise slot, under the line DFTTest(..), type
Decimate()
KEEP THIS FILTER INACTIVE

4)Right click, do JOIN ALL FILTERS

5) result:
- the CROP has disappeared :-(
- Field and Noise got merged that's OK
- Noise slot is still there !!! :-(

**Comment by Dendraspis** [comment]
That's the longest instruction I have ever read.  🤣   Gimme a minute ...

**Comment by tkozybski** [comment]
But I didn't do any joins and the last filter is just not visible, it's there in the main screen, also works on the yesterday's version. I tried adding more to see if settings will update but everything after that "limit" is just not visible.

I don't check it on a project, but on my default template.

Another thing I just noticed. In the main screen, when you have many filters and a scrollbar, if you scroll, the checkboxes will dissapear.

**Comment by Dendraspis** [comment]
@44vince44 
Could reproduce it - stupid mistake, but fixed:
[StaxRip v2.1.8.4D_CEscroll4.zip](https://github.com/staxrip/staxrip/files/6095397/StaxRip.v2.1.8.4D_CEscroll4.zip)

**Comment by 44vince44** [comment]
gimme 20 min ;-) i'll be back and check also the other things @tkozybski pointed out

**Comment by 44vince44** [comment]
OK here are some test results 😄 
1) the joining seems working correctly (i'll do some more tests later)
2) As @tkozybski said, if you have many filter slots, in the MAIN WINDOW you get a scroll bar. So scroll down, the checkboxes disappear completely
3) I can't reproduce the other issue of @tkozybski. I have 7 filter slots, filled them, they all appear ....
4) Now that would be very nice if the extra height that was introduced with this mod could be removed 😁 (I think I know the answer)

**Comment by Dendraspis** [comment]
Okay, seems like this is solved, because I also can not see any problem with the Code Editor now.

The Checkbox-thing is another issue that belongs to the last ListView issue, where the selected item used system colors. I'll try to fix it later...

**Comment by 44vince44** [comment]
@Dendraspis and the extra height that we need to remove.. really no solution for that? look how much it's taking. I'm sure there's something you can do by putting a condition related to the scrollbar.

![SNAG-0002](https://user-images.githubusercontent.com/17828567/110211433-49eb7680-7e9f-11eb-807c-864020758357.jpg)

**Comment by 44vince44** [comment]
@Dendraspis new bug
When you join **ALL FILTERS**

The lines that come from the inactive filters have the "comment" character # at the END of the line.
It should be at the beginning!

**Comment by Dendraspis** [comment]
Oh boy.  😁 

Okay, first the gap issue. I already optimized it so only almost one extra line is there and for me it looks:
<img src="https://user-images.githubusercontent.com/50659978/110212631-9c776380-7e9c-11eb-9492-edc3a4bccfe0.PNG">

But you may have a slightly older version without this fix.

**Comment by Dendraspis** [comment]
> @Dendraspis new bug
> When you join **ALL FILTERS**, the inactive filters that are joined get a comment # mark, that is GOOD, except it is not in the right place

This is a bug, that was there for several months.  🤣 
Seems nobody used this command so far, which I can understand, because it joined everything to *Source* which is not nice for further editing.  🤔 

But it is fixed.  😁

**Comment by 44vince44** [comment]
The way it is optimized in your case is very good (I see a difference when the slot has more than 2 written lines), so maybe it's ok!

**Comment by 44vince44** [comment]
OK THEN HOTFIX PLEASE !!!! 🤣

**Comment by Dendraspis** [comment]
😂  Check this out:
[StaxRip v2.1.8.4D_CEscroll5.zip](https://github.com/staxrip/staxrip/files/6095652/StaxRip.v2.1.8.4D_CEscroll5.zip)

**Comment by 44vince44** [comment]
Seems all ok here... we'll follow up in the discussions/highlight for further fixes regarding checkboxes , right ?

**Comment by Dendraspis** [comment]
Yeah, definitely not related to this issue.  😊 
So I close this now...

**Comment by tkozybski** [comment]
I know how to reproduce the missing filter issue, but it requires a lot of text in an filter. I have like 30-50 lines of comments about all parameters of the function, when I delete them, the missing filter appears. I just added one extra line to what you see on the left picture, resulting the right one.

[![scrollbar-issue-3.png](https://i.postimg.cc/1XdM5zYY/scrollbar-issue-3.png)](https://postimg.cc/wRhDw9zh)

That's why I asked for scrollbar for the code editor, like you have in main menu. If it's anything like in C#'s WPF, then it's totally doable by embedding a grid.  Monitor resolution wouldn't limit you on how many filters you could have and joining filters to save space wouldn't be needed.

**Comment by Dendraspis** [comment]
Is it possible, that there are empty lines in ***Misc/SelectRangeEvery*** ?
In ***Misc/SMDegrain Different*** the same.

As I already said, there is a point where you run out of space. The whole control disappears, because it cannot be painted within its limits.
The only solution code-wise would be decrease the number of lines per textbox from 15 to maybe 10?!
User-wise you can only merge two filters, so gerade it enough space, but sadly limited usability.

**Comment by tkozybski** [comment]
I tinkered with the code a bit and if you add this in CodeEditor.vb, scrolling is possible. But it requires more changes to look decent.
`MainFlowLayoutPanel.AutoScroll = True`

**Comment by 44vince44** [comment]
@Dendraspis if this works .... it's maybe better 😕

**Comment by tkozybski** [comment]
It's obviously way more work than just throwing this line, I just wanted to point that it's certainly possible with WinForms.

**Comment by Dendraspis** [comment]
Everything is possible if you take drawing everything by hand into consideration.  😁 
Of course this one line solves one problem, but also create some other. And as far I can tell it's not that easy to take care of the resize and flow order considering display resolution, scaling, ...

So let's keep it open and see, if anybody else needs it or tries to implement it.  

@tkozybski 
Feel free to implement a proper solution and create a Pull Request.  😉

**Comment by 44vince44** [comment]
@Dendraspis in the mean time, can you please take a look at this:

1) do some filter editing like you did here: https://github.com/staxrip/staxrip/issues/567#issuecomment-791976634
2) press OK to close filter editor
3) re-open filter editor

see how much wasted space there is....

**Comment by Dendraspis** [comment]
@44vince44 
What should I see??  😕 

***BEFORE pressing OK:***

<img width="441" src="https://user-images.githubusercontent.com/50659978/110217243-4bbf3500-7eb3-11eb-8e6d-b4898d7afc8e.PNG"> 

***AFTER pressing OK and reopening Code Editor:***

<img width="442" src="https://user-images.githubusercontent.com/50659978/110217245-4c57cb80-7eb3-11eb-8042-d59f1006aee6.PNG">

**Comment by 44vince44** [comment]
Well the pressing ok and reopening was only to ensure that you'll see the wasted space at maximum:

![sgan](https://user-images.githubusercontent.com/17828567/110217687-14a25100-7ebe-11eb-9397-a615d2c28067.jpg)

**Comment by Dendraspis** [comment]
There are empty lines - made by pressing *Return* after the last "Filters". If I would click there and press the *Reset* key. it would disappear.

You can also fill the whole control with empty lines by pressing *Enter* 20 times. Waste of space, but technically text.

**Comment by 44vince44** [comment]
@Dendraspis 

1) last empty line issue
> There are empty lines - made by pressing Return after the last "Filters"

If you delete the empty line after last line of text, then close with OK, then reopen Code Editor, **you'll find the empty line is back.**

2) there is an additional gap introduced with this mod, probably related to the minimum size of the scrollbar. But when the scrollbar is not there because only a few lines are typed, surely it can be removed, right ?

![ffffddd](https://user-images.githubusercontent.com/17828567/110226345-0f152d00-7ef7-11eb-9a60-c400ca1ab4e6.jpg)

**Comment by Dendraspis** [comment]
1. I'm not sure if I would change it, because this is very helpful if you wanna add new lines, especially when the last line is quite long.
But last time I tried I couldn't see such an behavior. 🤔

2. Nope, this is necessary for what I already mentioned: to prevent scrollbar flickering.
But maybe someone else can find a better solution. 🤔

**Comment by 44vince44** [comment]
> Nope, this is necessary for what I already mentioned: to prevent scrollbar flickering.
> But maybe someone else can find a better solution.

Yes I remember you said that at the beginning, and I was wishing it could be worked around. If it can't be, then no problem, we'll adapt.

**Comment by Dendraspis** [comment]
Maybe @stax76 or somebody else can find a better solution, but I fought enough battles in the last weeks with the UI.  🙄

**Comment by 44vince44** [comment]
>  I fought enough battles in the last weeks with the UI

I couldn't agree more 🤣

**Comment by 44vince44** [comment]
@Dendraspis Join filters is not working anymore (and there is no more Join all, join active etc...) that's maybe the only thing that was lost.
Tested in v6.

But the scrollbar is working!

**Comment by 44vince44** [comment]
This is all working OK after the minor incident @Dendraspis !

**Comment by stax76** [comment]
I've changed it a bit and hope it's OK.

It was necessary to reset the menu however.

**Comment by 44vince44** [comment]
@stax76 the changes you made with the size result in a better looking code Editor (no more extra unused space), but when you add lines, there is too much flicker with the scrollbar. Are you aware of that ?

**Comment by Dendraspis** [comment]
He made my height calculation undone.  😞 

@stax76 
Nothing wrong with it, as long there is another solution for the scrollbar flickering.  😁

**Comment by 44vince44** [comment]
Scrollbar flickering is very annoying :-(

**Comment by Dendraspis** [comment]
Oh yeah, so either increasing the height or removing the scrollbar.

But meanwhile I got used to the scrollbar and the fact, that it indicates ***Hey, there is more than you can see***.

**Comment by stax76** [comment]
You mean the scrollbar flashing when you add a new line by pressing enter? I understand why it happens, but fixing it might be tricky.

The scrollbars have to be enabled when the text control is initialized, changing it afterwards on demand causes an access violation. I also tried to enable horizontal scrollbar, that failed too. It's a native windows control, apparently somehow buggy.

**Comment by Dendraspis** [comment]
Indeed, that's why I increased the height a bit to not let the scrollbar appear before the height is adjusted.
I also tried setting scrollbars only if MaxTextHeight is reached, but this causes flickering, too.

So you know why I made it the way it was made.  😁 
So it seems we are back at either increasing the height or removing the scrollbar.  🤔

**Comment by 44vince44** [comment]
Now you are facing the same dilemma as yesterday, apparently 😁

**Comment by stax76** [comment]
Ok, now I understand why you changed the height. I've an idea, change height when the enter key is down, will try this.

**Comment by Dendraspis** [comment]
@44vince44 
There are enough flickering issues for everybody.   🤣 

@stax76 
Sometimes I make weird changes on purpose.  😉  😁 
That might work!  👍

**Comment by jlw4049** [comment]
You guys are machines lately
